### ISSUE TITLE
Paragraph separator character causes TS Server to get into bad state

### ISSUE THREAD

**Issue by mjbvz** [microsoft/TypeScript#38078]
From https://github.com/microsoft/vscode/issues/95685

**TypeScript Version:**  3.9.0-dev.20200420

**Search Terms:**
- invisible character
- paragraph
- diagnostics / errors

**Code**
For the JS
```ts
var ftext = [ "Und", "dann", "eines"];
var a = 1;
```

Note there is an invisible [paragraph separator](http://www.fileformat.info/info/unicode/char/2029/index.htm) character after the opening`[`

1. In the file, delete the space after the `=` on the second line.

**Bug:**
TS starts reporting bogus errors:

<img width="608" alt="Screen Shot 2020-04-20 at 3 31 24 PM" src="https://user-images.githubusercontent.com/12821956/79805885-fff7b180-831b-11ea-90e8-4b2566bf861c.png">

It looks the paragraph character messes up the parsing and causes the error line numbers to be off by one

```
[Trace  - 22:31:41.88] <semantic> Event received: semanticDiag (0).
Data: {
    "file": "/Users/matb/projects/san/test.js",
    "diagnostics": [
        {
            "start": {
                "line": 2,
                "offset": 8
            },
            "end": {
                "line": 2,
                "offset": 12
            },
            "text": "Cannot find name 'dann'.",
            "code": 2304,
            "category": "error"
        },
        {
            "start": {
                "line": 2,
                "offset": 16
            },
            "end": {
                "line": 2,
                "offset": 21
            },
            "text": "Cannot find name 'eines'.",
            "code": 2304,
            "category": "error"
        }
    ]
}
```

**Comment by DanielRosenwasser** [comment]
[In JavaScript, Paragraph Separator is a line terminator](https://tc39.es/ecma262/#sec-line-terminators); however, in the edits described above, VS Code says that edits are occur on line 2. 

From TypeScript's perspective, line 2 occurs immediately after the Paragraph Separator. Ultimately, editors need to account for each language's newline settings if they intent to use line/offset encodings for commands, so I don't think we can fix this on the TypeScript side.

**Comment by DanielRosenwasser** [comment]
See also:

* the original issue filed over 5 years ago https://github.com/microsoft/TypeScript/issues/2343
* the issue tracked on the LSP side https://github.com/microsoft/language-server-protocol/issues/80 (but that's been closed)

**Comment by mjbvz** [comment]
Here's the edit we send to TSServer when deleting the space after the `=` in `a = 1`:

```
[Trace  - 23:20:02.22] <syntax> Sending request: updateOpen (12). Response expected: yes. Current queue length: 0
Arguments: {
    "changedFiles": [
        {
            "fileName": "/Users/matb/projects/san/test.ts",
            "textChanges": [
                {
                    "newText": "",
                    "start": {
                        "line": 2,
                        "offset": 8
                    },
                    "end": {
                        "line": 2,
                        "offset": 9
                    }
                }
            ]
        }
    ],
    "closedFiles": [],
    "openFiles": []
```

The indexes one based so they look correct to me. 

@RyanCavanaugh @DanielRosenwasser Does this also match what you see?

**Comment by DanielRosenwasser** [comment]
Yeah, that's what I'm seeing as well, but if I understand correctly the edit needs to happen at line 3 because the paragraph separator is a line terminator.

**Comment by mjbvz** [comment]
VS Code does not consider the character a line break. Looks like VS does though

@alexdima should be able to say if this is intentional or not. If you have this character in your code and it's not inside a string though, it's probably a bug

**Comment by alexdima** [comment]
Currently, VS Code considers the following characters/sequences to be line terminators: `CRLF`, `CR` or `LF`.

I feel that this is a difficult topic to get wide agreement on, since we need to work with multiple programming languages, each one with its own definition of what a line terminator is. So, if a compiler creates a diagnostic at line 100, it appears that line number is defined according to the programming language's own specification.

From my reading of https://www.unicode.org/reports/tr14/tr14-32.html , Unicode defines a mandatory line break after the following:
* **LB4**. Always break after hard line breaks (`BK`). In the `BK` class there are the following: 
  * `000C` - FORM FEED (FF)
  * `000B` - LINE TABULATION (VT) - **[OPTIONAL]**
  * `2028` - LINE SEPARATOR (LS)
  * `2029` - PARAGRAPH SEPARATOR (PS)
* **LB5**. Treat `CR` followed by `LF`, as well as `CR`, `LF`, and `NL` as hard line breaks:
  * `000D`x`000A`- CARRIAGE RETURN (CR) x LINE FEED (LF)
  * `000D` - CARRIAGE RETURN (CR)
  * `000A` - LINE FEED (LF)
  * `0085` - NEXT LINE (NEL)

Here I'm looking at the most popular languages used in VS Code for which I could find a specification or where the decision is not pushed out to specific implementations (looking at you, C++).

It looks like [the JS spec Table 33](https://www.ecma-international.org/publications/files/ECMA-ST/ECMA-262.pdf) defines the following as line terminators. I suppose TS uses the same:
![image](https://user-images.githubusercontent.com/5047891/80284276-e8566980-871d-11ea-9ecd-4cfc9a614137.png)

Interestingly, [the C# spec A.1.1 Line terminators](https://www.ecma-international.org/publications/files/ECMA-ST-ARCH/ECMA-334%201st%20edition%20December%202001.pdf) defines the same line terminators:
![image](https://user-images.githubusercontent.com/5047891/80284297-f906df80-871d-11ea-9b82-26510d289fc9.png)

[HTML](https://html.spec.whatwg.org/#newlines) however defines newlines as just `CR`, `LF` and `CRLF`:
![image](https://user-images.githubusercontent.com/5047891/80284373-79c5db80-871e-11ea-9013-beb3e272fec6.png)

[Python](https://docs.python.org/3/reference/lexical_analysis.html#physical-lines) also defines just `CR`, `LF` and `CRLF`:
![image](https://user-images.githubusercontent.com/5047891/80284529-6404e600-871f-11ea-9c38-be59d12c749f.png)

[PHP](https://github.com/php/php-langspec/blob/fe55bee01293a925120cfe330b62a94fc5140e35/spec/09-lexical-structure.md#comments) also defines just `CR`, `LF` and `CRLF`:
![image](https://user-images.githubusercontent.com/5047891/80284334-38cdc700-871e-11ea-8091-6480740c71d9.png)

[Java](https://docs.oracle.com/javase/specs/jls/se7/html/jls-3.html#jls-3.4) defines `LF`, `CR` and `CRLF`:
![image](https://user-images.githubusercontent.com/5047891/80284786-e04bf900-8720-11ea-8a72-56be5fd9b054.png)

[YAML section 3.1.4 Line Breaks](https://yaml.org/spec/history/2002-10-31.html) uses the following:
![image](https://user-images.githubusercontent.com/5047891/80284740-906d3200-8720-11ea-9d42-b36646fa8b12.png)

To sum up:

| char/seq | Unicode | JS | C# | HTML | Python | PHP | Java | YAML |
|----------|---------|----|----|------|--------|-----|------|------|
|   CRLF   |   ✅   | ✅ | ✅ |  ✅  |   ✅  |  ✅ |  ✅  | ✅  |
|    CR    |   ✅   | ✅ | ✅ |  ✅  |   ✅  |  ✅ |  ✅  | ✅  |
|    LF    |   ✅   | ✅ | ✅ |  ✅  |   ✅  |  ✅ |  ✅  | ✅  |
|    LS    |   ✅   | ✅ | ✅ |      |        |     |      | ✅  |
|    PS    |   ✅   | ✅ | ✅ |      |        |     |      | ✅  |
|    NEL   |   ✅   |    |    |       |       |     |      |  ✅  |
|    FF    |   ✅   |    |    |       |       |     |      |      |
|    VT    |   ✅   |    |    |       |       |     |      |      |

I am not sure what would be best, since it would be very difficult to make VS Code end-of-line sequence dependent on the file's language and changing things now might break language extensions which follow our definition. 

@dbaeumer Is the end-of-line sequence something that is specified in the Language Server Protocol?

**Comment by alexdima** [comment]
Actually, a straight forward way to tackle this would be for us to prompt users and ask for permission to "fix" their files. Most likely, `LS`, `PS` or `NEL` are inserted by accident in files by copy-pasting and they are unwanted there anyways. I've opened https://github.com/microsoft/vscode/issues/96142 to track this.

**Comment by dbaeumer** [comment]
@alexdima yes it is specified here: https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocuments

**Comment by dbaeumer** [comment]
The LSP support `\r`, `\n` and `\r\n`
